File: LeetCode_Algorithm/93(Restore IP Addresses)/93.js
var restoreIpAddresses = function(s) {
  // sol1 66ms 89% 42.4MB 85%
  ans = []
  solve(s, [])
  return ans
};
var solve = (s, temp) => {
  if (temp.length == 4) {
    let string = temp.join("")
    string = string.slice(0, string.length-1)
    ans.push(string)
    return ;
  }
  let i = 1;
  while (i < 4 && i <= s.length) {
    let num = s.slice(0, i)
    if (num.length > 1 && num[0] == '0') {
    i += 1; continue
    }
    if (0 <= parseInt(num) && parseInt(num) <= 255) {
    num += '.'
    temp.push(num)
    solve(s.slice(i, ), temp)
    temp.pop()
    }
    i += 1
    }
}

File: LeetCode_Algorithm/93(Restore IP Addresses)/93.py
'''
LeetCode 93 : Restore IP Addresses
숫자로 이루어진 문자열이 주어질 때 가능한 IP를 모두 리턴하라.

# sol1 35ms 86% 14MB 31%
재귀 함수
'''
class Solution:
    def restoreIpAddresses(self, s: str) -> list[str]:
        self.ans = []
        self.solve(s, [])
        return self.ans

    def solve(self, s, temp):
        if len(temp) == 4:
            if len(s) > 0:
                return ;
            string = "".join(temp);
            string = string[0:len(string)-1]
            self.ans.append(string)
            return ;
        i = 1
        while i < 4:
            if i <= len(s):
                if 0 <= int(s[:i]) <= 255:
                    if len(s[:i]) > 1 and s[0] == '0':
                        i += 1
                        continue
                    temp.append(s[:i] + '.')
                    self.solve(s[i:], temp)
                    temp.pop()
            i += 1

print(Solution().restoreIpAddresses("25525511135"))
            

            